tests/abi/abi.c

```diff
@@ -0,0 +1,175 @@
+/*
+ * This file is distributed under the MIT License. See LICENSE.md for details.
+ */
+
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "for_each.h"
+#include "functions.h"
+
+#ifdef __x86_64__
+#include "platform_specific/x86_64.h"
+#elif __x86__
+#include "platform_specific/x86.h"
+#else
+_Static_assert(0, "unsupported platform architecture");
+#endif
+
+/* LFSR generator */
+
+static uint64_t LFSR = 0xdc984ef2647af904ull;
+uint64_t getNextLFSR(void) {
+  uint32_t Iteration;
+  for (Iteration = 0; Iteration < 64; ++Iteration)
+    LFSR = (LFSR >> 1) | ((((LFSR >> 1) ^ (LFSR >> 2)) & 1) << 63);
+  return LFSR;
+}
+
+/* Helpers */
+
+#define GET_1ST_HELPER(THIS_ONE, ...) THIS_ONE
+#define GET_1ST(THIS_ONE, ...) GET_1ST_HELPER(THIS_ONE, __VA_ARGS__)
+#define FIRST_REGISTER GET_1ST(REGISTER_LIST)
+
+#define STACK_VALUE_COUNT 32
+#define REGISTER_COUNT COUNT_ARGUMENTS(REGISTER_LIST)
+#define GENERATED_COUNT REGISTER_COUNT + STACK_VALUE_COUNT
+
+void generate(REGISTER_C_TYPE *Generated) {
+  for (uint32_t Counter = 0; Counter < GENERATED_COUNT; ++Counter)
+    Generated[Counter] = getNextLFSR();
+}
+
+void dump(REGISTER_C_TYPE *Generated) {
+  /* Dump registers */
+  puts("  Registers:");
+#define DUMP_REGISTER(REGISTER)                           \
+  do {                                                    \
+    printf("    - Name: \"%s\"\n      Value: 0x%016lx\n", \
+           STRINGIFY(REGISTER),                           \
+           Generated[RegisterIndex--]);                   \
+  } while (0);
+  {
+    uint32_t RegisterIndex = REGISTER_COUNT - 1;
+    FOR_EACH(DUMP_REGISTER, REGISTER_LIST);
+  }
+#undef DUMP_REGISTER
```

Comment:
Unless you see a benefit in using macros so aggresively, I suggest to convert `REGISTER_LIST` to a `const char *registers_list[] = { "rax", "rbx", ... };` and drop the related macros.

Reply:
Wait, I'm pretty sure those are only used at compile time, so run-time array doesn't cut it :(
Even more so considering I'm iterating on this list at runtime.

The only possible thing I can see us being able to improve here is going to a macro list of strings, e.g.
```cpp
#define REGISTER_LIST "rax", "rbx", "rcx", [...]
```
but I'd rather use an extra `STRINGIFY` than deal with potential complications of a misplaced `"`, etc.
As a bonus `STRINGIFY` is a lot more explicit about what's actually going in the mess down below.

Reply:
> are only used at compile time

What do you mean by that?
```c
const char *registers_list[] = { "rax", "rbx", ... };

void dump_registers() {
  const size_t register_count = elementsof(registers_list);
  uint32_t RegisterIndex = register_count - 1;
  for (unsigned I = 0; I < register_count; ++I) {
     printf("    - Name: \"%s\"\n      Value: 0x%016lx\n",
           registers_list[I],
           Generated[RegisterIndex--]);
  }
}
```

Reply:
And compile time loops are important since we can't exactly tell the compiler to skip one register to use for looping, etc. I feel a lot safer if every single loop is already unrolled before the compiler even touches it.

Reply:
I'm not opposed, but it's more of a consistency issue at that point. If we're already using this uber-ugly (but more optimized and unavoidable elsewhere) solution, we might as well keep this one avoidable point the same just to keep things consistent. I'm fine either way, though.

Reply:
I agree, but let's push to the minimum the macros. They spread like XML.